File: apps/server1/src/api/components/user/user-service.ts
import httpStatus from 'http-status'

import { ServerError } from '../../../server-error'
import { IUserRepository } from './user-repository'
import { User } from './user-types'

interface IUserService {
  registerUser(user: User): Promise<User>
  retrieveAndCountUsers(page: number, limit: number): Promise<[User[], number]>
  retrieveUser(userId: string): Promise<User>
  replaceUser(userId: string, user: User): Promise<User>
  removeUser(userId: string): Promise<User>
}

class UserService implements IUserService {
  userRepository: IUserRepository

  constructor(userRepository: IUserRepository) {
    this.userRepository = userRepository
  }

  async registerUser(user: User): Promise<User> {
    let newUser: User | undefined

    try {
      newUser = await this.userRepository.createUser(user)
    } catch (error) {
      const message = 'An error occurred when creating a user'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: user,
        cause: error,
      })
    }

    if (newUser === undefined) {
      const message = 'User could not be created'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: user,
        cause: undefined,
      })
    }

    return newUser
  }

  async retrieveAndCountUsers(
    page: number,
    limit: number,
  ): Promise<[User[], number]> {
    let users: User[] | undefined
    let total: number | undefined

    try {
      ;[users, total] = await this.userRepository.readAndCountUsers(page, limit)
    } catch (error) {
      const message = 'An error occurred when reading and counting users'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: { page: page, limit: limit },
        cause: error,
      })
    }

    if (users === undefined || total === undefined) {
      const message = 'Users could not be read and counted'
      throw new ServerError(message, httpStatus.NOT_FOUND, {
        context: { page: page, limit: limit },
        cause: undefined,
      })
    }

    return [users, total]
  }

  async retrieveUser(userId: string): Promise<User> {
    let user: User | undefined

    try {
      user = await this.userRepository.readUser(userId)
    } catch (error) {
      const message = 'An error occurred when reading a user'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: userId,
        cause: error,
      })
    }

    if (user === undefined) {
      const message = 'User could not be read'
      throw new ServerError(message, httpStatus.NOT_FOUND, {
        context: userId,
        cause: undefined,
      })
    }

    return user
  }

  async replaceUser(userId: string, user: User): Promise<User> {
    let updatedUser: User | undefined

    try {
      updatedUser = await this.userRepository.updateUser(userId, user)
    } catch (error) {
      const message = 'An error occurred when updating a user'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: { userId: userId, user: user },
        cause: error,
      })
    }

    if (updatedUser === undefined) {
      const message = 'User could not be updated'
      throw new ServerError(message, httpStatus.NOT_FOUND, {
        context: { userId: userId, user: user },
        cause: undefined,
      })
    }

    return updatedUser
  }

  async removeUser(userId: string): Promise<User> {
    let user: User | undefined

    try {
      user = await this.userRepository.deleteUser(userId)
    } catch (error) {
      const message = 'An error occurred when deleting a user'
      throw new ServerError(message, httpStatus.INTERNAL_SERVER_ERROR, {
        context: userId,
        cause: error,
      })
    }

    if (user === undefined) {
      const message = 'User could not be removed'
      throw new ServerError(message, httpStatus.NOT_FOUND, {
        context: userId,
        cause: undefined,
      })
    }

    return user
  }
}

export { IUserService, UserService }

File: apps/server2/src/api/components/user/user_service.py
from abc import ABC, abstractmethod

from fastapi import status

from api.components.user.user_models import User
from api.components.user.user_repository import UserRepository
from server_error import Detail, ServerError


class IUserService(ABC):
    @abstractmethod
    async def register_user(self, user: User) -> User:
        raise Exception("NotImplementedException")

    @abstractmethod
    async def retrieve_and_count_users(
        self, page: int, limit: int
    ) -> tuple[list[User], int]:
        raise Exception("NotImplementedException")

    @abstractmethod
    async def retrieve_user(self, user_id: str) -> User:
        raise Exception("NotImplementedException")

    @abstractmethod
    async def replace_user(self, user_id: str, user: User) -> User:
        raise Exception("NotImplementedException")

    @abstractmethod
    async def remove_user(self, user_id: str) -> User:
        raise Exception("NotImplementedException")


class UserService(IUserService):
    def __init__(self, user_repository: UserRepository):
        self.user_repository = user_repository

    async def register_user(self, user: User) -> User:
        new_user: User | None

        try:
            new_user = await self.user_repository.create_user(user)
        except Exception as error:
            message = "An error occurred when creating a user"
            print(message, error)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context=user, cause=error),
            )

        if new_user is None:
            message = "User could not be created"
            print(message)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context=user, cause=None),
            )

        return new_user

    async def retrieve_and_count_users(
        self, page: int, limit: int
    ) -> tuple[list[User], int]:
        records: list[User] | None
        total: int | None

        try:
            records, total = await self.user_repository.read_and_count_users(
                page, limit
            )
        except Exception as error:
            message = "An error occurred when reading and counting users"
            print(message, error)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context={"page": page, "limit": limit}, cause=str(error)),
            )

        if records is None or total is None:
            message = "Users could not be read and counted"
            print(message)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context={"page": page, "limit": limit}, cause=None),
            )

        return records, total

    async def retrieve_user(self, user_id: str) -> User:
        user: User | None

        try:
            user = await self.user_repository.read_user(user_id)
        except Exception as error:
            message = "An error occurred when reading a user"
            print(message, error)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context=user_id, cause=str(error)),
            )

        if user is None:
            message = "User could not be read"
            print(message)
            raise ServerError(
                message,
                status.HTTP_404_NOT_FOUND,
                Detail(context=user_id, cause=None),
            )

        return user

    async def replace_user(self, user_id: str, user: User) -> User:
        user: User | None

        try:
            user = await self.user_repository.update_user(user_id, user)
        except Exception as error:
            message = "An error occurred when updating a user"
            print(message, error)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context={"user_id": user_id, "user": user}, cause=str(error)),
            )

        if user is None:
            message = "User could not be updated"
            print(message)
            raise ServerError(
                message,
                status.HTTP_404_NOT_FOUND,
                Detail(context={"user_id": user_id, "user": user}, cause=None),
            )

        return user

    async def remove_user(self, user_id: str) -> User:
        user: User | None

        try:
            user = await self.user_repository.delete_user(user_id)
        except Exception as error:
            message = "An error occurred when deleting a user"
            print(message, error)
            raise ServerError(
                message,
                status.HTTP_500_INTERNAL_SERVER_ERROR,
                Detail(context=user_id, cause=str(error)),
            )

        if user is None:
            message = "User could not be removed"
            print(message)
            raise ServerError(
                message,
                status.HTTP_404_NOT_FOUND,
                Detail(context=user_id, cause=None),
            )

        return user
